storagenode/pieces/readwrite.go

```diff
@@ -0,0 +1,122 @@
+// Copyright (C) 2019 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package pieces
+
+import (
+	"bufio"
+	"hash"
+	"io"
+
+	"github.com/zeebo/errs"
+
+	"storj.io/storj/pkg/pkcrypto"
+	"storj.io/storj/storage"
+
+	_ "storj.io/storj/storage/filestore"
+)
+
+// Writer implements a piece writer that writes content to blob store and calculates a hash.
+type Writer struct {
+	buf  bufio.Writer
+	hash hash.Hash
+	blob storage.BlobWriter
+	size int64
+}
+
+// NewWriter creates a new writer for storage.BlobWriter.
+func NewWriter(blob storage.BlobWriter, bufferSize int) (*Writer, error) {
+	w := &Writer{}
+	w.buf = *bufio.NewWriterSize(blob, bufferSize)
+	w.blob = blob
+	w.hash = pkcrypto.NewHash()
+	return w, nil
+}
+
+// Write writes data to the blob and calculates the hash.
+func (w *Writer) Write(data []byte) (int, error) {
+	_, err := w.hash.Write(data)
```

Comment:
Why are we adding the data to the hash first? Should we add only those bytes that were _actually_ written to buf _after_ writing them successfully?

Reply:
yes, fixed